### PR TITLE
Support passing layer instances to produce attentional hidden states

### LINKED ISSUE #14972: Support optional activation function to produce attentional hidden states
*This is a feature request.*

When using an `AttentionWrapper` with the `attention_layer_size` argument set, the cell output and the context are combined and fed through a dense layer to produce an attentional hidden state.

However, it is also common to apply an activation function on this layer output (typically *tanh*). This is for example described in:

> Minh-Thang Luong, Hieu Pham, Christopher D. Manning. "Effective Approaches to Attention-based Neural Machine Translation." EMNLP 2015.  https://arxiv.org/abs/1508.04025

I would be happy to contribute. Do we want to add a new argument `attention_layer_activation_fn=None` to the `AttentionWrapper` constructor?

### PR DESCRIPTION
**Pull request by guillaumekln**
This PR closes #14972.

### CODE REVIEW

**Comment by drpngx** [comment]
@ebrevdo WDYT?

File: tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py [hunk 1]
```diff
@@ -1144,13 +1149,28 @@ def __init__(self,
...
+      if len(attention_layer_activations) != len(attention_layer_sizes):
+        raise ValueError(
+            "If provided, attention_layer_activation must contain exactly one "
+            "activation function per attention layer, saw: %d vs %d"
+            % (len(attention_layer_activations), len(attention_layer_sizes)))
+      attention_layers_param = zip(
+          attention_layer_sizes, attention_layer_activations)
       self._attention_layers = tuple(
           layers_core.Dense(
-              attention_layer_size,
```
Comment (ebrevdo) [inline tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: I wonder if it makes sense to just allow a list of attention_layers as an alternative to passing attention_layer_size.  Then users could pass their own Dense (or something else) here.  what do you think?

Comment (guillaumekln) [reply tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: That makes sense. Do you still want to keep compatibility with the existing behavior? It seems cleaner and more powerful to only allow passing a list of `Layers`. I suppose you suggest to add a new option and raise an exception when both are set.

**Comment by ebrevdo** [comment]
Backwards compatibility with the simpler parameter would be great.

**Comment by guillaumekln** [comment]
When directly passing layers, the output size is not known until `build(input_shape)` is called on the layer instance. Some parts of `AttentionWrapper` assume that the output size is defined at construction time which would not be the case here. Any thoughts?

**Comment by ebrevdo** [comment]
The `Dense` layer has a special hidden function, [_compute_output_shape](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/core.py#L169), which can be called before `build`.  Theoretically, all layer objects should have it (in practice, any other Layers may have to add it separately).  We use this function in `tf.contrib.seq2seq` to infer the output static shape.

**Comment by guillaumekln** [comment]
Thanks, I used this method in the latest commit. Please have a second look.

**Comment by rmlarsen** [comment]
@ebrevdo can you take another look, please?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by ebrevdo — CHANGES_REQUESTED** [review]
Add checks that layer_sizes is none != layers is None

So the user can't pass both

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by guillaumekln** [comment]
Hi @ebrevdo, the check "layer_sizes is none != layers is None" is already in the code. See https://github.com/tensorflow/tensorflow/pull/14974/files#diff-fb6bbbf892e4a4722fe40b8abd0b60bfR1194.

I also noticed that `compute_output_shape` is now public. The PR is updated accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by guillaumekln** [comment]
Sorry for bumping this thread. Is there something else I need to do?

File: tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py [hunk 2]
```diff
@@ -1077,6 +1077,7 @@ def __init__(self,
                cell,
                attention_mechanism,
                attention_layer_size=None,
+               attention_layer=None,
```
Comment (ebrevdo) [inline tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: Can you move this towards the end of the function signature to avoid breaking existing users?

File: tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py [hunk 3]
```diff
@@ -1182,6 +1191,10 @@ def __init__(self,
             "cell_input_fn must be callable, saw type: %s"
             % type(cell_input_fn).__name__)
 
+    if attention_layer_size is not None and attention_layer is not None:
```
Comment (ebrevdo) [inline tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: If ... is not None == ... is not None:
  Really

Comment (ebrevdo) [reply tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: Err.. then Raise

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py [hunk 4]
```diff
@@ -1124,7 +1125,14 @@ def __init__(self,
...
         feed the context and cell output into the attention layer to generate
         attention at each time step. If attention_mechanism is a list,
-        attention_layer_size must be a list of the same length.
+        attention_layer_size must be a list of the same length. If
+        attention_layer is set, this must be None.
+      attention_layer: A list of `tf.layers.Layer` instances or a
+        single `tf.layers.Layer` instance taking the context and cell output as
+        inputs to generate attention at each time step. If None (default), use
+        the context as attention at each time step. If attention_mechanism is a
+        list, attention_layer must be a list of the same length. If
```
Comment (ebrevdo) [inline tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: What happens if mechanisms is not a list but this is?

Comment (guillaumekln) [reply tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: `attention_mechanism` is converted to a `tuple` in the constructor when it is not a sequence. The opposite is also true.

File: tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py (hunk 3 above)
Comment (guillaumekln) [reply tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: I think the condition is correct here. Additionally, both can be `None` which is a valid case.

File: tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py [hunk 5]
```diff
@@ -1200,6 +1213,22 @@ def __init__(self,
...
+          else (attention_layer,))
+      if len(self._attention_layers) != len(attention_mechanisms):
+        raise ValueError(
+            "If provided, attention_layer must contain exactly one "
+            "layer per attention_mechanism, saw: %d vs %d"
+            % (len(self._attention_layers), len(attention_mechanisms)))
+      self._attention_layer_size = sum(
+          layer.compute_output_shape(
+              [None,
+               cell.output_size + mechanism.values.shape[-1].value])[-1].value
```
Comment (ebrevdo) [inline tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py]: do yo need to confirm that mechanism.values.shape[-1].value is not None and that the result of compute_output_shape [-1].value is not None?

**Review by ebrevdo — APPROVED** [review]
generally LGTM; thank you for the contribution!
